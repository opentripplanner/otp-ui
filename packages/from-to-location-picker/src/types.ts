// eslint-disable-next-line prettier/prettier
import type { Location } from "@opentripplanner/types";

export type FromToPickerProps = {
  /**
   * Specifies the label to be rendered, or if set to true, renders the default label "Plan a trip:".
   */
  label?: boolean | React.ReactElement | string;
  /**
   * A specific location to associate with this. This is only used when combined
   * with the setLocation prop.
   */
  location?: Location;
  /**
   * Triggered when the user clicks on the "from" button.
   */
  onFromClick?: () => void;
  /**
   * Triggered when the user clicks on the "to" button.
   */
  onToClick?: () => void;
  /**
   * Triggered when the user clicks either the "from" or "to" button and there
   * are no from/to specific handler functions defined as props.
   *
   * Passes an argument as follows:
   * { locationType: "from/to", location, reverseGeocode: false }
   */
  setLocation?: (_: {
    locationType: string,
    // eslint-disable-next-line @typescript-eslint/no-shadow
    location: Location,
    reverseGeocode: boolean
  }) => void;
  /**
   * Determines whether icons are shown on the "from" and "to" buttons.
   */
  showIcons?: boolean;
};
