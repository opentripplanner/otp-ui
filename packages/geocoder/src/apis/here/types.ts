// Prettier does not support typescript annotation
// eslint-disable-next-line prettier/prettier
import type { LonLatInput } from "@conveyal/lonlat"

export interface HereResponse {
  items: Item[];
  point?: LonLatInput; // used in reverse response
}

export type Rect = {
  maxLat: number;
  maxLon: number;
  minLat: number;
  minLon: number;
};

export type Boundary = {
  country?: string;
  rect?: Rect;
};

export interface Item {
  access: Position[];
  address: Address;
  categories: Category[];
  distance: number;
  foodTypes: Category[];
  id: string;
  position: Position;
  resultType: string;
  scoring: Scoring;
  title: string;
}

export interface Position {
  lat: number;
  lng: number;
}

export interface Address {
  city: string;
  countryCode: string;
  countryName: string;
  county: string;
  district: string;
  houseNumber: string;
  label: string;
  postalCode: string;
  state: string;
  stateCode: string;
  street: string;
}

export interface Category {
  id: string;
  name: string;
  primary?: boolean;
}

export interface Scoring {
  fieldScore: FieldScore;
  queryScore: number;
}

export interface FieldScore {
  district: number;
  placeName: number;
}
