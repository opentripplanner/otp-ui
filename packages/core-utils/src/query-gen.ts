import { LonLatOutput } from "@conveyal/lonlat";
import { print } from "graphql";
import {
  ModeSetting,
  ModeSettingValues,
  TransportMode
} from "@opentripplanner/types";

import DefaultPlanQuery from "./planQuery.graphql";

type InputBanned = {
  routes?: string;
  agencies?: string;
  trips?: string;
  stops?: string;
  stopsHard?: string;
};

type InputPreferred = {
  routes?: string;
  agencies?: string;
  unpreferredCost?: string;
};

type OTPQueryParams = {
  arriveBy: boolean;
  date?: string;
  from: LonLatOutput & { name?: string };
  modes: TransportMode[];
  modeSettings: ModeSetting[];
  time?: string;
  numItineraries?: number;
  to: LonLatOutput & { name?: string };
  banned?: InputBanned;
  preferred?: InputPreferred;
};

type GraphQLQuery = {
  query: string;
  variables: Record<string, unknown>;
};

/**
 * Mode Settings can contain additional modes to add to the query,
 * this function extracts those additional modes from the settings
 * and returns them in an array.
 * @param modeSettings List of mode settings with values populated
 * @returns Additional transport modes to add to query
 */
export function extractAdditionalModes(
  modeSettings: ModeSetting[],
  enabledModes: TransportMode[]
): TransportMode[] {
  return modeSettings.reduce<TransportMode[]>((prev, cur) => {
    // First, ensure that the mode associated with this setting is even enabled
    if (!enabledModes.map(m => m.mode).includes(cur.applicableMode)) {
      return prev;
    }

    // In checkboxes, mode must be enabled and have a transport mode in it
    if (
      (cur.type === "CHECKBOX" || cur.type === "SUBMODE") &&
      cur.addTransportMode &&
      cur.value
    ) {
      return [...prev, cur.addTransportMode];
    }
    if (cur.type === "DROPDOWN") {
      const transportMode = cur.options.find(o => o.value === cur.value)
        ?.addTransportMode;
      if (transportMode) {
        return [...prev, transportMode];
      }
    }
    return prev;
  }, []);
}

/**
 * Generates every possible mathematical subset of the input TransportModes.
 * Uses code from:
 * https://stackoverflow.com/questions/5752002/find-all-possible-subset-combos-in-an-array
 * @param array Array of input transport modes
 * @returns 2D array representing every possible subset of transport modes from input
 */
function combinations(array: TransportMode[]): TransportMode[][] {
  if (!array) return [];
  return (
    // eslint-disable-next-line no-bitwise
    new Array(1 << array.length)
      .fill(null)
      // eslint-disable-next-line no-bitwise
      .map((e1, i) => array.filter((e2, j) => i & (1 << j)))
  );
}

/**
 * This constant maps all the transport mode to a broader mode type,
 * which is used to determine the valid combinations of modes used in query generation.
 */
export const SIMPLIFICATIONS = {
  AIRPLANE: "TRANSIT",
  BICYCLE: "PERSONAL",
  BUS: "TRANSIT",
  CABLE_CAR: "TRANSIT",
  CAR: "CAR",
  FERRY: "TRANSIT",
  FLEX: "SHARED", // TODO: this allows FLEX+WALK. Is this reasonable?
  FUNICULAR: "TRANSIT",
  GONDOLA: "TRANSIT",
  RAIL: "TRANSIT",
  SCOOTER: "PERSONAL",
  SUBWAY: "TRANSIT",
  TRAM: "TRANSIT",
  TRANSIT: "TRANSIT",
  WALK: "WALK"
};

// Inclusion of "TRANSIT" alone automatically implies "WALK" in OTP
const VALID_COMBOS = [
  ["WALK"],
  ["PERSONAL"],
  ["TRANSIT", "SHARED"],
  ["WALK", "SHARED"],
  ["TRANSIT"],
  ["TRANSIT", "PERSONAL"],
  ["TRANSIT", "CAR"]
];

const BANNED_TOGETHER = ["SCOOTER", "BICYCLE", "CAR"];

export const TRANSIT_SUBMODES = Object.keys(SIMPLIFICATIONS).filter(
  mode => SIMPLIFICATIONS[mode] === "TRANSIT" && mode !== "TRANSIT"
);
export const TRANSIT_SUBMODES_AND_TRANSIT = Object.keys(SIMPLIFICATIONS).filter(
  mode => SIMPLIFICATIONS[mode] === "TRANSIT"
);

function isCombinationValid(
  combo: TransportMode[],
  queryTransitSubmodes: string[]
): boolean {
  if (combo.length === 0) return false;

  // All current qualifiers currently simplify to "SHARED"
  const simplifiedModes = Array.from(
    new Set(combo.map(c => (c.qualifier ? "SHARED" : SIMPLIFICATIONS[c.mode])))
  );

  // Ensure that if we have one transit mode, then we include ALL transit modes
  if (simplifiedModes.includes("TRANSIT")) {
    // Don't allow TRANSIT along with any other submodes
    if (queryTransitSubmodes.length && combo.find(c => c.mode === "TRANSIT")) {
      return false;
    }

    if (
      combo.reduce((prev, cur) => {
        if (queryTransitSubmodes.includes(cur.mode)) {
          return prev - 1;
        }
        return prev;
      }, queryTransitSubmodes.length) !== 0
    ) {
      return false;
    }
    // Continue to the other checks
  }

  // OTP doesn't support multiple non-walk modes
  if (BANNED_TOGETHER.filter(m => combo.find(c => c.mode === m)).length > 1) {
    return false;
  }

  return !!VALID_COMBOS.find(
    vc =>
      simplifiedModes.every(m => vc.includes(m)) &&
      vc.every(m => simplifiedModes.includes(m))
  );
}

/**
 * Generates a list of queries for OTP to get a comprehensive
 * set of results based on the modes input.
 * @param params OTP Query Params
 * @returns Set of parameters to generate queries
 */
export function generateCombinations(params: OTPQueryParams): OTPQueryParams[] {
  const completeModeList = [
    ...extractAdditionalModes(params.modeSettings, params.modes),
    ...params.modes
  ];

  // List of the transit *submodes* that are included in the input params
  const queryTransitSubmodes = completeModeList
    .filter(mode => TRANSIT_SUBMODES.includes(mode.mode))
    .map(mode => mode.mode);

  return combinations(completeModeList)
    .filter(combo => isCombinationValid(combo, queryTransitSubmodes))
    .map(combo => ({ ...params, modes: combo }));
}

/**
 * Generates a query for OTP GraphQL API based on parameters.
 * @param param0 OTP2 Parameters for the query
 * @param planQuery Override the default query for OTP
 * @returns A fully formed query+variables ready to be sent to GraphQL backend
 */
export function generateOtp2Query(
  {
    arriveBy,
    banned,
    date,
    from,
    modes,
    modeSettings,
    numItineraries,
    preferred,
    time,
    to
  }: OTPQueryParams,
  planQuery = DefaultPlanQuery
): GraphQLQuery {
  // This extracts the values from the mode settings to key value pairs
  const modeSettingValues = modeSettings.reduce((prev, cur) => {
    if (cur.type === "SLIDER" && cur.inverseKey) {
      prev[cur.inverseKey] = cur.high - cur.value + cur.low;
    }
    prev[cur.key] = cur.value;

    // If we assign a value on true, return the value (or null) instead of a boolean.
    if (cur.type === "CHECKBOX" && cur.truthValue) {
      prev[cur.key] = cur.value === true ? cur.truthValue : null;
    }
    return prev;
  }, {}) as ModeSettingValues;

  const {
    bikeReluctance,
    carReluctance,
    walkSpeed,
    walkReluctance,
    wheelchair
  } = modeSettingValues;

  return {
    query: print(planQuery),
    variables: {
      arriveBy,
      banned,
      bikeReluctance,
      carReluctance,
      date,
      fromPlace: `${from.name}::${from.lat},${from.lon}}`,
      modes,
      numItineraries,
      preferred,
      time,
      toPlace: `${to.name}::${to.lat},${to.lon}}`,
      walkReluctance,
      walkSpeed,
      wheelchair
    }
  };
}
