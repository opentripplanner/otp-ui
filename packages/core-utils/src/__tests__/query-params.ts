import { ModeSetting, TransportMode } from "@opentripplanner/types";

import { reduceOtpFlexModes } from "../query";
import queryParams, { getCustomQueryParams } from "../query-params";
import { extractAdditionalModes, generateCombinations } from "../query-gen";

const customWalkDistanceOptions = [
  {
    text: "200 m",
    value: 200
  },
  {
    text: "500 m",
    value: 500
  }
];

function modeStrToTransportMode(m): TransportMode {
  const splitVals = m.split("_");
  return {
    mode: splitVals[0],
    qualifier: splitVals?.[1] || null
  };
}

const mockLatLon = {
  lat: 1,
  lon: 2
};

function expectModes(modes: string[], expectedModes: string[][]) {
  const generatedModesList = generateCombinations({
    from: mockLatLon,
    modes: modes.map(modeStrToTransportMode),
    modeSettings: [],
    to: mockLatLon
  });
  const expandedExpectedModesList = expectedModes.map(em => ({
    from: mockLatLon,
    modes: em.map(modeStrToTransportMode),
    modeSettings: [],
    to: mockLatLon
  }));
  return it(
    modes.join(" "),
    () =>
      expect(generatedModesList.length === expandedExpectedModesList.length) &&
      expect(new Set(generatedModesList)).toEqual(
        new Set(expandedExpectedModesList)
      )
  );
}

describe("extract-modes", () => {
  const mode = {
    mode: "UNICYCLE"
  };

  const testTransportMode: TransportMode = {
    mode: "testMode"
  };

  const checkboxModeSetting: ModeSetting = {
    addTransportMode: mode,
    applicableMode: testTransportMode.mode,
    icon: null,
    key: "test",
    label: "test",
    type: "CHECKBOX",
    value: true
  };

  const dropdownModeSetting: ModeSetting = {
    applicableMode: testTransportMode.mode,
    key: "test",
    label: "test",
    options: [{ text: "testOption", value: "1", addTransportMode: mode }],
    type: "DROPDOWN",
    value: "1"
  };

  it("determines whether a checkbox setting is extracted correctly", () => {
    expect(
      extractAdditionalModes([checkboxModeSetting], [testTransportMode])
    ).toEqual([mode]);
  });
  it("determines whether a dropdown setting is extracted correctly", () => {
    expect(
      extractAdditionalModes([dropdownModeSetting], [testTransportMode])
    ).toEqual([mode]);
  });
  it("determines whether a checkbox setting set to false is ignored", () => {
    expect(
      extractAdditionalModes(
        [{ ...checkboxModeSetting, value: false }],
        [testTransportMode]
      )
    ).toEqual([]);
  });
  it("determines whether a checkbox setting with no modes is ignored", () => {
    expect(extractAdditionalModes([{ ...checkboxModeSetting }], [])).toEqual(
      []
    );
  });
});

describe("query-gen", () => {
  describe("generateCombinations", () => {
    expectModes(["WALK"], [["WALK"]]);
    expectModes(["WALK", "TRANSIT"], [["WALK"], ["TRANSIT"]]);
    expectModes(
      ["WALK", "TRANSIT", "BICYCLE"],
      [["WALK"], ["TRANSIT"], ["BICYCLE"], ["TRANSIT", "BICYCLE"]]
    );
    expectModes(
      ["WALK", "TRANSIT", "CAR"],
      [["WALK"], ["TRANSIT"], ["TRANSIT", "CAR"]]
    );
    expectModes(["TRANSIT", "CAR"], [["TRANSIT"], ["TRANSIT", "CAR"]]);
    expectModes(["CAR"], []);
    expectModes(
      ["WALK", "TRANSIT", "BICYCLE", "CAR"],
      [
        ["WALK"],
        ["TRANSIT"],
        ["TRANSIT", "BICYCLE"],
        ["BICYCLE"],
        ["TRANSIT", "CAR"]
      ]
    );
    expectModes(
      ["BICYCLE_RENT", "TRANSIT", "WALK"],
      [
        ["TRANSIT"],
        ["BICYCLE_RENT", "TRANSIT"],
        ["BICYCLE_RENT", "WALK"],
        ["WALK"]
      ]
    );
    expectModes(
      ["BICYCLE_RENT", "BICYCLE", "TRANSIT", "WALK"],
      [
        ["TRANSIT"],
        ["BICYCLE_RENT", "TRANSIT"],
        ["BICYCLE", "TRANSIT"],
        ["BICYCLE_RENT", "WALK"],
        ["BICYCLE"],
        ["WALK"]
      ]
    );
    expectModes(
      ["BICYCLE_RENT", "BICYCLE", "WALK"],
      [["BICYCLE_RENT", "WALK"], ["BICYCLE"], ["WALK"]]
    );
    expectModes(
      ["SCOOTER_RENT", "BICYCLE_RENT", "TRANSIT", "WALK"],
      [
        ["TRANSIT"],
        ["BICYCLE_RENT", "TRANSIT"],
        ["BICYCLE_RENT", "WALK"],
        ["SCOOTER_RENT", "TRANSIT"],
        ["SCOOTER_RENT", "WALK"],
        ["WALK"]
      ]
    );

    expectModes(
      ["CAR_HAIL", "TRANSIT"],
      [["TRANSIT"], ["CAR_HAIL", "TRANSIT"]]
    );
    expectModes(
      ["CAR_HAIL", "BICYCLE_RENT", "TRANSIT"],
      [["TRANSIT"], ["CAR_HAIL", "TRANSIT"], ["BICYCLE_RENT", "TRANSIT"]]
    );
    expectModes(
      ["CAR_HAIL", "BICYCLE_RENT", "TRANSIT", "WALK"],
      [
        ["TRANSIT"],
        ["CAR_HAIL", "TRANSIT"],
        ["CAR_HAIL", "WALK"],
        ["BICYCLE_RENT", "TRANSIT"],
        ["BICYCLE_RENT", "WALK"],
        ["WALK"]
      ]
    );
    expectModes(
      ["FLEX", "TRANSIT", "WALK"],
      [["TRANSIT"], ["FLEX", "TRANSIT"], ["FLEX", "WALK"], ["WALK"]]
    );
    expectModes(
      ["FLEX", "SCOOTER_RENT", "TRANSIT", "WALK"],
      [
        ["TRANSIT"],
        ["FLEX", "TRANSIT"],
        ["WALK"],
        ["FLEX", "WALK"],
        ["FLEX", "SCOOTER_RENT", "WALK"], // Is this sensible?
        ["FLEX", "SCOOTER_RENT", "TRANSIT"],
        ["SCOOTER_RENT", "WALK"],
        ["SCOOTER_RENT", "TRANSIT"]
      ]
    );
    expectModes(
      ["FLEX", "SCOOTER_RENT", "TRANSIT"],
      [
        ["TRANSIT"],
        ["FLEX", "TRANSIT"],
        ["FLEX", "SCOOTER_RENT", "TRANSIT"],
        ["SCOOTER_RENT", "TRANSIT"]
      ]
    );
    expectModes(
      // Transit is required to enable other transit submodes
      ["BUS", "RAIL", "GONDOLA", "TRAM", "TRANSIT"],
      [["BUS", "RAIL", "GONDOLA", "TRAM"]]
    );
    expectModes(
      // Transit is required to enable other transit submodes
      ["TRANSIT"],
      [["TRANSIT"]]
    );
  });
});

describe("query-params", () => {
  describe("getCustomQueryParams", () => {
    it("should return the original unmodified queryParams if no customizations", () => {
      expect(getCustomQueryParams()).toEqual(queryParams);
    });

    it("should return queryParams with customizations", () => {
      const customizations = {
        maxWalkDistance: {
          label: "Max Walk Distance In Meters",
          options: customWalkDistanceOptions
        }
      };
      expect(getCustomQueryParams(customizations)).toMatchSnapshot();
    });

    it("should ignore unknown query params", () => {
      const customizations = {
        unknownQueryParameter: {
          label: "Unknown query parameter",
          options: customWalkDistanceOptions
        }
      };
      expect(getCustomQueryParams(customizations)).toEqual(queryParams);
    });
  });
});

describe("flex-reducer", () => {
  it("should not touch a query that doesn't include flex modes", () => {
    expect(reduceOtpFlexModes(["WALK", "TRANSIT", "BIKE"])).toStrictEqual([
      "WALK",
      "TRANSIT",
      "BIKE"
    ]);
  });
  it("should modify a query that includes some flex modes", () => {
    expect(
      reduceOtpFlexModes(["WALK", "TRANSIT", "BIKE", "FLEX_DIRECT"])
    ).toStrictEqual(["WALK", "TRANSIT", "BIKE", "FLEX"]);
  });
  it("should modify a query that includes all flex modes", () => {
    expect(
      reduceOtpFlexModes([
        "WALK",
        "TRANSIT",
        "BIKE",
        "FLEX_DIRECT",
        "FLEX_ACCESS",
        "FLEX_EGRESS"
      ])
    ).toStrictEqual(["WALK", "TRANSIT", "BIKE", "FLEX"]);
    expect(
      reduceOtpFlexModes([
        "FLEX_DIRECT",
        "BIKE",
        "FLEX_ACCESS",
        "WALK",
        "FLEX_EGRESS",
        "TRANSIT"
      ])
    ).toStrictEqual(["FLEX", "BIKE", "WALK", "TRANSIT"]);
  });
  it("should modify a query that includes only flex modes", () => {
    expect(
      reduceOtpFlexModes(["FLEX_DIRECT", "FLEX_ACCESS", "FLEX_EGRESS"])
    ).toStrictEqual(["FLEX"]);
  });
  it("should modify a query that includes duplicate flex modes", () => {
    expect(
      reduceOtpFlexModes([
        "FLEX_DIRECT",
        "FLEX_DIRECT",
        "FLEX_ACCESS",
        "FLEX_EGRESS"
      ])
    ).toStrictEqual(["FLEX"]);
  });
});
