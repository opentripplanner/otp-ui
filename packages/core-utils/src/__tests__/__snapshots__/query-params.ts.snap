// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query-params getCustomQueryParams should return queryParams with customizations 1`] = `
Array [
  Object {
    "default": null,
    "itineraryRewrite": [Function],
    "name": "from",
    "profileRewrite": [Function],
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
  },
  Object {
    "default": null,
    "itineraryRewrite": [Function],
    "name": "to",
    "profileRewrite": [Function],
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
  },
  Object {
    "default": [Function],
    "name": "date",
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
  },
  Object {
    "default": [Function],
    "name": "time",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "default": "NOW",
    "itineraryRewrite": [Function],
    "name": "departArrive",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "default": "07:00",
    "name": "startTime",
    "routingTypes": Array [
      "PROFILE",
    ],
  },
  Object {
    "default": "09:00",
    "name": "endTime",
    "routingTypes": Array [
      "PROFILE",
    ],
  },
  Object {
    "default": "WALK,TRANSIT",
    "name": "mode",
    "profileRewrite": [Function],
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
  },
  Object {
    "default": true,
    "name": "showIntermediateStops",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "applicable": [Function],
    "default": 1609,
    "label": "Max Walk Distance In Meters",
    "name": "maxWalkDistance",
    "options": Array [
      Object {
        "text": "200 m",
        "value": 200,
      },
      Object {
        "text": "500 m",
        "value": 500,
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": 4828,
    "itineraryRewrite": [Function],
    "label": "Maximum Bike",
    "name": "maxBikeDistance",
    "options": Array [
      Object {
        "text": "1/4 mile",
        "value": 402.3,
      },
      Object {
        "text": "1/2 mile",
        "value": 804.7,
      },
      Object {
        "text": "3/4 mile",
        "value": 1207,
      },
      Object {
        "text": "1 mile",
        "value": 1609,
      },
      Object {
        "text": "2 miles",
        "value": 3219,
      },
      Object {
        "text": "3 miles",
        "value": 4828,
      },
      Object {
        "text": "5 miles",
        "value": 8047,
      },
      Object {
        "text": "10 miles",
        "value": 16093,
      },
      Object {
        "text": "20 miles",
        "value": 32187,
      },
      Object {
        "text": "30 miles",
        "value": 48280,
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": "QUICK",
    "label": "Optimize for",
    "name": "optimize",
    "options": Array [
      Object {
        "text": "Speed",
        "value": "QUICK",
      },
      Object {
        "text": "Fewest Transfers",
        "value": "TRANSFERS",
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": "SAFE",
    "itineraryRewrite": [Function],
    "label": "Optimize for",
    "name": "optimizeBike",
    "options": [Function],
    "routingTypes": Array [
      "ITINERARY",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": 15,
    "label": "Max Walk Time",
    "name": "maxWalkTime",
    "options": Array [
      Object {
        "text": "5 minutes",
        "value": 5,
      },
      Object {
        "text": "10 minutes",
        "value": 10,
      },
      Object {
        "text": "15 minutes",
        "value": 15,
      },
      Object {
        "text": "20 minutes",
        "value": 20,
      },
      Object {
        "text": "30 minutes",
        "value": 30,
      },
      Object {
        "text": "45 minutes",
        "value": 45,
      },
      Object {
        "text": "1 hour",
        "value": 60,
      },
    ],
    "routingTypes": Array [
      "PROFILE",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "high": 10,
    "label": "walk reluctance",
    "labelHigh": "More Transit",
    "labelLow": "More Walking",
    "low": 1,
    "name": "walkReluctance",
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
    "selector": "SLIDER",
    "step": 0.5,
  },
  Object {
    "applicable": [Function],
    "default": 20,
    "label": "Max Bike Time",
    "name": "maxBikeTime",
    "options": Array [
      Object {
        "text": "5 minutes",
        "value": 5,
      },
      Object {
        "text": "10 minutes",
        "value": 10,
      },
      Object {
        "text": "15 minutes",
        "value": 15,
      },
      Object {
        "text": "20 minutes",
        "value": 20,
      },
      Object {
        "text": "30 minutes",
        "value": 30,
      },
      Object {
        "text": "45 minutes",
        "value": 45,
      },
      Object {
        "text": "1 hour",
        "value": 60,
      },
    ],
    "routingTypes": Array [
      "PROFILE",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": 3.58,
    "label": "Bicycle Speed",
    "name": "bikeSpeed",
    "options": Array [
      Object {
        "text": "6 MPH",
        "value": 2.68,
      },
      Object {
        "text": "8 MPH",
        "value": 3.58,
      },
      Object {
        "text": "10 MPH",
        "value": 4.47,
      },
      Object {
        "text": "12 MPH",
        "value": 5.36,
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": 4828,
    "itineraryRewrite": [Function],
    "label": "Maximum E-scooter Distance",
    "name": "maxEScooterDistance",
    "options": Array [
      Object {
        "text": "1/4 mile",
        "value": 402.3,
      },
      Object {
        "text": "1/2 mile",
        "value": 804.7,
      },
      Object {
        "text": "3/4 mile",
        "value": 1207,
      },
      Object {
        "text": "1 mile",
        "value": 1609,
      },
      Object {
        "text": "2 miles",
        "value": 3219,
      },
      Object {
        "text": "3 miles",
        "value": 4828,
      },
      Object {
        "text": "5 miles",
        "value": 8047,
      },
      Object {
        "text": "10 miles",
        "value": 16093,
      },
      Object {
        "text": "20 miles",
        "value": 32187,
      },
      Object {
        "text": "30 miles",
        "value": 48280,
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "applicable": [Function],
    "default": 250,
    "itineraryRewrite": [Function],
    "label": "E-scooter Power",
    "name": "watts",
    "options": Array [
      Object {
        "text": "Kid's hoverboard (6mph)",
        "value": 125,
      },
      Object {
        "text": "Entry-level scooter (11mph)",
        "value": 250,
      },
      Object {
        "text": "Robust E-scooter (18mph)",
        "value": 500,
      },
      Object {
        "text": "Powerful E-scooter (24mph)",
        "value": 1500,
      },
    ],
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
    "selector": "DROPDOWN",
  },
  Object {
    "default": false,
    "name": "ignoreRealtimeUpdates",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "companies",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "applicable": [Function],
    "default": false,
    "icon": <ForwardRef(Wheelchair) />,
    "label": "Prefer Wheelchair Accessible Routes",
    "name": "wheelchair",
    "routingTypes": Array [
      "ITINERARY",
      "PROFILE",
    ],
    "selector": "CHECKBOX",
  },
  Object {
    "name": "bannedRoutes",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "default": 3,
    "name": "numItineraries",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "default": Array [],
    "itineraryRewrite": [Function],
    "name": "intermediatePlaces",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "default": 900,
    "name": "otherThanPreferredRoutesPenalty",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "preferredRoutes",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "maxPreTransitTime",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "waitReluctance",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "driveDistanceReluctance",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "driveTimeReluctance",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "waitAtBeginningFactor",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bikeSwitchTime",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bikeSwitchCost",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "minTransferTime",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "preferredAgencies",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "unpreferredRoutes",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "unpreferredAgencies",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "walkBoardCost",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bikeBoardCost",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "whiteListedRoutes",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bannedAgencies",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "whiteListedAgencies",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bannedTrips",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bannedStops",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "bannedStopsHard",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "transferPenalty",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "nonpreferredTransferPenalty",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "maxTransfers",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "batch",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "startTransitStopId",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "startTransitTripId",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "clampInitialWait",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "reverseOptimizeOnTheFly",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "boardSlack",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "alightSlack",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "locale",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "disableRemainingWeightHeuristic",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "flexFlagStopBufferSize",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "flexUseReservationServices",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "flexUseEligibilityServices",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "flexIgnoreDrtAdvanceBookMin",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "maxHours",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "useRequestedDateTimeInMaxHours",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "disableAlertFiltering",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "geoidElevation",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "invalidDateStrategy",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "minTransitDistance",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "searchTimeout",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "pathComparator",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "onlyTransitTrips",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
  Object {
    "name": "minimumMicromobilitySpeed",
    "routingTypes": Array [
      "ITINERARY",
    ],
  },
]
`;
